src/test/ram_expression_equal_clone_test.cpp

```diff
@@ -0,0 +1,196 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2019, The Souffle Developers. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file ram_expression_equal_clone_test.cpp
+ *
+ * Tests equal and clone function of RamExpression classes.
+ *
+ ***********************************************************************/
+
+#include "RamExpression.h"
+
+#include "test.h"
+
+namespace souffle {
+
+namespace test {
+
+TEST(RamIntrinsicOperator, CloneAndEquals) {
+    // ADD(number(1), number(2))
+    std::vector<std::unique_ptr<RamExpression>> a_args;
+    a_args.emplace_back(new RamNumber(1));
+    a_args.emplace_back(new RamNumber(2));
+    RamIntrinsicOperator a(FunctorOp::ADD, std::move(a_args));
+
+    std::vector<std::unique_ptr<RamExpression>> b_args;
+    b_args.emplace_back(new RamNumber(1));
+    b_args.emplace_back(new RamNumber(2));
+    RamIntrinsicOperator b(FunctorOp::ADD, std::move(b_args));
+    EXPECT_EQ(a, b);
+    EXPECT_NE(&a, &b);
+
+    RamIntrinsicOperator* c = a.clone();
+    EXPECT_EQ(a, *c);
+    EXPECT_NE(&a, c);
+    delete c;
+
+    // NEG(number(1))
+    std::vector<std::unique_ptr<RamExpression>> d_args;
+    d_args.emplace_back(new RamNumber(1));
+    RamIntrinsicOperator d(FunctorOp::NEG, std::move(d_args));
+
+    std::vector<std::unique_ptr<RamExpression>> e_args;
+    e_args.emplace_back(new RamNumber(1));
+    RamIntrinsicOperator e(FunctorOp::NEG, std::move(e_args));
+    EXPECT_EQ(d, e);
+    EXPECT_NE(&d, &e);
+
+    RamIntrinsicOperator* f = d.clone();
+    EXPECT_EQ(d, *f);
+    EXPECT_NE(&d, f);
+    delete f;
+}
+
+TEST(RamUserDefinedOperator, CloneAndEquals) {
+    // define binary functor NE check if two RamExpressions are not equal
+    // NE(number(1), number(10))
+    std::vector<std::unique_ptr<RamExpression>> a_args;
+    a_args.emplace_back(new RamNumber(1));
+    a_args.emplace_back(new RamNumber(10));
+    RamUserDefinedOperator a("NE", "Binary", std::move(a_args));
+
+    std::vector<std::unique_ptr<RamExpression>> b_args;
+    b_args.emplace_back(new RamNumber(1));
+    b_args.emplace_back(new RamNumber(10));
+    RamUserDefinedOperator b("NE", "Binary", std::move(b_args));
+    EXPECT_EQ(a, b);
+    EXPECT_NE(&a, &b);
+
+    RamUserDefinedOperator* c = a.clone();
+    EXPECT_EQ(a, *c);
+    EXPECT_NE(&a, c);
+    delete c;
+}
+
+TEST(RamTupleElement, CloneAndEquals) {
+    // t0.1
```

Comment:
I'm pretty sure this comment is useful if you have an in-depth knowledge of the Ram* and their print output, but it would be better if it were easier to understand.